### PR TITLE
PR: Allow creation of temporary credentials

### PR DESCRIPTION
**Pull request by martindurant**
This is to enable passing around credentials inside distributed
without using the raw key/secret from the user

### CODE REVIEW

File: s3fs/core.py [hunk 1]
```diff
@@ -137,16 +139,39 @@ def connect(self, refresh=False):
...
+        if self.anon:
+            return {'anon': True}
+        if self.token:  # already has temporary cred
+            return {'key': self.key, 'secret': self.secret, 'token': self.token,
+                    'anon': False}
+        sts = boto3.Session(self.key, self.secret, self.token,
+                            **self.kwargs).client('sts')
+        cred = sts.get_session_token(DurationSeconds=3600)['Credentials']
+        return {'key': cred['AccessKeyId'], 'secret': cred['SecretAccessKey'],
+                'token': cred['SessionToken'], 'anon': False}
```
Comment (mrocklin) [inline s3fs/core.py]: Should we cache the `token` or `secret` onto self after calling this?

Comment (martindurant) [inline s3fs/core.py]: Plausibly, but perhaps we want new temporary credentials every time, and of course they are designed to expire (so we would have to at least check for expiry).

Comment (mrocklin) [inline s3fs/core.py]: OK, presumably we call this infrequently anyway, right?  (i.e., less than once a second)

Comment (martindurant) [inline s3fs/core.py]: Should be called just once in the main thread when launching a parallel distributed s3 job. An alternative to call even less frequently is to set `s3delegated = S3FileSystems(**s3. get_delegated_s3pars)` and then only use s3delegated.

**Comment by mrocklin** [comment]
Should we merge?  Distributed tests are failing without this.

**Pull request merged**